client/src/main/java/org/asynchttpclient/util/AuthenticatorUtils.java

```diff
@@ -81,25 +89,307 @@ private static String computeDigestAuthentication(Realm realm, Uri uri) {
             append(builder, "cnonce", realm.getCnonce(), true);
         }
+        // RFC7616: userhash parameter (optional, not implemented yet)
         builder.setLength(builder.length() - 2); // remove tailing ", "
+        Charset wireCs = (realm.getCharset() == StandardCharsets.UTF_8)
+                ? StandardCharsets.UTF_8
+                : ISO_8859_1;
+        return new String(StringUtils.charSequence2Bytes(builder, wireCs), wireCs);
+    }
+
+    /**
+     * Calculates the digest response value for HTTP Digest Authentication.
+     * This method computes HA1 and HA2 (including entity-body hash for auth-int).
+     *
+     * @param realm   The authentication realm containing credentials and challenge info
+     * @param request The HTTP request (needed for method, uri, and body)
+     * @return The computed response hex string
+     * @throws UnsupportedOperationException if qop=auth-int but body cannot be hashed
+     */
+    static String computeDigestResponse(Realm realm, Request request) {
+        String algorithm = realm.getAlgorithm() != null ? realm.getAlgorithm() : "MD5";
+        String qop = realm.getQop() != null ? realm.getQop() : "auth";
+
+        String hashAlgorithm = algorithm.replace("-sess", "");
+        Charset wireCharset = realm.getCharset() != null ?
+                realm.getCharset() : StandardCharsets.ISO_8859_1;
+
+        // Calculate HA1
+        String ha1 = calculateHA1(realm, algorithm);
+
+        // Get request URI
+        Uri uri = request.getUri();
+        String requestUri = uri.getPath() +
+                (uri.getQuery() != null ? "?" + uri.getQuery() : "");
+
+        // Calculate HA2
+        String ha2;
+        if ("auth-int".equals(qop)) {
+            String bodyHash = computeBodyHash(request, realm);
+            ha2 = calculateHA2AuthInt(request, requestUri, bodyHash, hashAlgorithm, wireCharset);
+        } else {
+            // Regular auth: HA2 = H(method:uri)
+            String a2Plain = request.getMethod() + ":" + requestUri;
+            MessageDigest md = MessageDigestUtils.pooledMessageDigest(hashAlgorithm);
+            try {
+                md.update(a2Plain.getBytes(wireCharset));
+                ha2 = MessageDigestUtils.bytesToHex(md.digest());
+            } finally {
+                md.reset();
+            }
+        }
+
+        // Build final response
+        String nc = realm.getNc() != null ? realm.getNc() : "00000001";
+        String cnonce = realm.getCnonce();
+        String nonce = realm.getNonce();
+
+        // response = H(HA1:nonce:nc:cnonce:qop:HA2)
+        String responseInput = ha1 + ":" + nonce + ":" + nc + ":" +
+                cnonce + ":" + qop + ":" + ha2;
+
+        MessageDigest md = MessageDigestUtils.pooledMessageDigest(hashAlgorithm);
+        try {
+            md.update(responseInput.getBytes(StandardCharsets.ISO_8859_1));
+            return MessageDigestUtils.bytesToHex(md.digest());
+        } finally {
+            md.reset();
```

Comment:
We don't need to reset `MessageDigest` since we will return the hash and it won't be reused anymore.